File: daily_problems/2026/08/0831/personal_submission/cf102129d_yawn_sean.cpp
#include <bits/stdc++.h>
// #include "atcoder/all"
// #pragma GCC optimize("Ofast")
// #pragma GCC optimize("unroll-loops")
// #pragma GCC target("sse,sse2,sse3,ssse3,sse4,popcnt,abm,mmx,avx2,tune=native")

using namespace std;

long long quickPow(long long base, long long power, long long mod) {
	if (power == 0) return 1 % mod;
	long long cur = quickPow(base, power / 2, mod);
	return power & 1 ? base * cur % mod * cur % mod : cur * cur % mod; 
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int k, mod = 1e9 + 7;
	cin >> k;

	vector<int> v1(k), v2(k);
	for (auto &x: v1) cin >> x;
	for (auto &x: v2) cin >> x;

	vector<vector<int>> xn(2 * k, vector<int>(k, 0));

	for (int i = 0; i < k; i ++) xn[i][i] = 1;

	for (int i = k; i < 2 * k; i ++) {
		for (int j = 0; j < k; j ++) {
			for (int x = 0; x < k; x ++) {
				xn[i][x] = (xn[i][x] + 1ll * xn[i - j - 1][x] * v1[j]) % mod;
			}
		}
	}

	auto solve = [&] (auto &self, int x) -> vector<int> {
		if (x < 2 * k) return xn[x];

		auto v = self(self, x / 2);
		int tag = x % 2;

		vector<int> res(2 * k, 0);

		for (int i = 0; i < k; i ++) {
			for (int j = 0; j < k; j ++) {
				res[i + j + tag] = (res[i + j + tag] + 1ll * v[i] * v[j]) % mod;
			}
		}

		vector<int> ans(k, 0);

		for (int i = 0; i < 2 * k; i ++) {
			for (int j = 0; j < k; j ++) {
				ans[j] = (ans[j] + 1ll * res[i] * xn[i][j]) % mod;
			}
		}

		return ans;
	};

	vector<vector<int>> equations(k, vector<int>(k + 1));

	for (int i = 0; i < k; i ++) {
		auto tmp = solve(solve, v2[k - 1] - v2[i]);
		for (int j = 0; j < k; j ++) {
			equations[j][i] = tmp[j];
		}
	}

	{
		auto tmp = solve(solve, v2[k - 1]);
		for (int j = 0; j < k; j ++) {
			equations[j][k] = tmp[j];
		}
	}

	int pt = 0;

	for (int i = 0; i < k; i ++) {
		int chosen = -1;

		while (pt < k + 1 && chosen == -1) {
			for (int j = i; j < k; j ++) {
				if (equations[i][pt]) {
					chosen = j;
					break;
				}
			}
			pt ++;
		}

		pt --;

		if (chosen != i) swap(equations[i], equations[chosen]);
		if (pt == k + 1) break;

		int v = quickPow(equations[i][pt], mod - 2, mod);

		for (int j = pt; j <= k; j ++) {
			equations[i][j] = 1ll * equations[i][j] * v % mod;
		}

		for (int j = i + 1; j < k; j ++) {
			int w = (mod - equations[j][pt]) % mod;
			for (int x = pt; x <= k; x ++) {
				equations[j][x] = (equations[j][x] + 1ll * equations[i][x] * w) % mod;
			}
		}
	}

	for (int i = k - 1; i >= 0; i --) {
		for (int j = 0; j <= k; j ++) {
			if (equations[i][j]) {
				for (int x = 0; x < i; x ++) {
					int w = (mod - equations[x][j]) % mod;
					for (int y = j; y <= k; y ++) {
						equations[x][y] = (equations[x][y] + 1ll * equations[i][y] * w) % mod;
					}
				}
				break;
			}
		}
	}

	for (int i = 0; i < k; i ++) {
		cout << equations[i][k] << " \n"[i == k - 1];
	}

	return 0;
}

File: daily_problems/2026/08/0831/personal_submission/cf102129d_yawn_sean.py
# Submission link: https://codeforces.com/gym/102129/submission/388650906
def main():
    k = II()
    v1 = LII()
    v2 = LII()
    
    mod = 10 ** 9 + 7
    
    xn = [[0] * k for _ in range(2 * k)]
    
    for i in range(k):
        xn[i][i] = 1
    
    for i in range(k, 2 * k):
        for j in range(k):
            for x in range(k):
                xn[i][x] += xn[i - j - 1][x] * v1[j]
                xn[i][x] %= mod
    
    def solve(x):
        if x < 2 * k: return xn[x]
        
        v = solve(x // 2)
        tag = x % 2
        
        res = [0] * (2 * k)
        
        for i in range(k):
            for j in range(k):
                res[i + j + tag] += v[i] * v[j]
                res[i + j + tag] %= mod
        
        ans = [0] * k
        
        for i in range(2 * k):
            for j in range(k):
                ans[j] += res[i] * xn[i][j]
                ans[j] %= mod
        
        return ans
    
    equations = [[0] * (k + 1) for _ in range(k)]
    
    for i in range(k):
        tmp = solve(v2[k - 1] - v2[i])
        for j in range(k):
            equations[j][i] = tmp[j]
    
    tmp = solve(v2[k - 1])
    for j in range(k):
        equations[j][k] = tmp[j]
    
    pt = 0
    
    for i in range(k):
        chosen = -1
        while pt < k + 1 and chosen == -1:
            for j in range(i, k):
                if equations[j][pt]:
                    chosen = j
                    break
            pt += 1
        
        pt -= 1
        
        if chosen != i:
            equations[i], equations[chosen] = equations[chosen], equations[i]
        
        if pt == k + 1: break
        
        v = pow(equations[i][pt], -1, mod)
        
        for j in range(pt, k + 1):
            equations[i][j] = equations[i][j] * v % mod
        
        for j in range(i + 1, k):
            w = (-equations[j][pt]) % mod
            for x in range(pt, k + 1):
                equations[j][x] += equations[i][x] * w
                equations[j][x] %= mod
    
    for i in range(k - 1, -1, -1):
        for j in range(k + 1):
            if equations[i][j]:
                
                for x in range(i):
                    w = (-equations[x][j]) % mod
                    for y in range(j, k + 1):
                        equations[x][y] += equations[i][y] * w
                        equations[x][y] %= mod
                
                break
    
    print(' '.join(str(equations[i][k]) for i in range(k)))